hazelcast-client/src/main/java/com/hazelcast/client/spi/impl/ClientInvocation.java

```diff
@@ -0,0 +1,260 @@
+package com.hazelcast.client.spi.impl;
+
+import com.hazelcast.client.HazelcastClientNotActiveException;
+import com.hazelcast.client.config.ClientProperties;
+import com.hazelcast.client.connection.nio.ClientConnection;
+import com.hazelcast.client.impl.HazelcastClientInstanceImpl;
+import com.hazelcast.client.impl.client.ClientRequest;
+import com.hazelcast.client.impl.client.RetryableRequest;
+import com.hazelcast.client.spi.ClientContext;
+import com.hazelcast.client.spi.ClientExecutionService;
+import com.hazelcast.client.spi.ClientInvocationService;
+import com.hazelcast.client.spi.EventHandler;
+import com.hazelcast.core.HazelcastInstanceNotActiveException;
+import com.hazelcast.core.LifecycleService;
+import com.hazelcast.executor.impl.client.RefreshableRequest;
+import com.hazelcast.nio.Address;
+import com.hazelcast.nio.Connection;
+import com.hazelcast.spi.exception.RetryableHazelcastException;
+
+import java.io.IOException;
+import java.util.concurrent.RejectedExecutionException;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import static com.hazelcast.client.config.ClientProperties.PROP_HEARTBEAT_INTERVAL_DEFAULT;
+
+public class ClientInvocation implements Runnable {
+
+    protected static final long RETRY_WAIT_TIME_IN_SECONDS = 1;
+    private final LifecycleService lifecycleService;
+    private final ClientInvocationService invocationService;
+    private final ClientExecutionService executionService;
+    private final ClientListenerServiceImpl listenerService;
+    private final ClientRequest request;
+    private final EventHandler handler;
+    private final long retryCountLimit;
+
+    private final ClientInvocationFuture clientInvocationFuture;
+    private final int heartBeatInterval;
+    private final AtomicInteger reSendCount = new AtomicInteger();
+
+    private boolean isBindToSingleConnection;
+    private Address address;
+    private int partitionId = -1;
```

Comment:
I prefer `address` and `partitionId` to be `final` fields. Are they set after invocation is constructed?